src/Features/LanguageServer/Protocol/LspServices/AbstractExportLspServiceAttribute.cs

```diff
@@ -0,0 +1,81 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Composition;
+using System.Linq;
+using Microsoft.CommonLanguageServerProtocol.Framework;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.LanguageServer;
+
+internal abstract class AbstractExportLspServiceAttribute : ExportAttribute
+{
+    /// <summary>
+    /// The full assembly-qualified type name of the service being exported.
+    /// </summary>
+    public string TypeName { get; }
+
+    /// <summary>
+    /// The LSP server for which this service applies to.  If null, this service applies to any server
+    /// with the matching contract name.
+    /// </summary>
+    public WellKnownLspServerKinds? ServerKind { get; }
+
+    /// <summary>
+    /// Services MEF exported as <see cref="ILspService"/> must by definition be stateless as they are
+    /// shared amongst all LSP server instances through restarts.
+    /// </summary>
+    public bool IsStateless { get; }
+
+    /// <summary>
+    /// The full assembly-qualified type names of the interfaces the service implements.
+    /// </summary>
+    public string[] InterfaceNames { get; }
+
+    private readonly Lazy<string?[]>? _lazyHandlerMethodData;
+
+    /// <summary>
+    /// If this this service implements <see cref="IMethodHandler"/>, returns a blob of binary data
+    /// that encodes an array of <see cref="HandlerMethodDetails"/>s; otherwise <see langword="null"/>.
+    /// </summary>
+    public string?[]? HandlerMethodData => _lazyHandlerMethodData?.Value;
+
+    protected AbstractExportLspServiceAttribute(
+        Type serviceType, string contractName, Type contractType, bool isStateless, WellKnownLspServerKinds serverKind)
+        : base(contractName, contractType)
+    {
+        Contract.ThrowIfFalse(serviceType.GetInterfaces().Contains(typeof(ILspService)), $"{serviceType.Name} does not inherit from {nameof(ILspService)}");
+        Contract.ThrowIfNull(serviceType.AssemblyQualifiedName);
+
+        TypeName = serviceType.AssemblyQualifiedName;
+        IsStateless = isStateless;
+        ServerKind = serverKind;
+
+        InterfaceNames = Array.ConvertAll(serviceType.GetInterfaces(), t => t.AssemblyQualifiedName!);
+
+        _lazyHandlerMethodData = typeof(IMethodHandler).IsAssignableFrom(serviceType)
+            ? new(() => CreateHandlerMethodData(serviceType))
+            : null;
+    }
+
+    private static string?[] CreateHandlerMethodData(Type handlerType)
+    {
+        var handlerMethods = HandlerMethodDetails.From(handlerType);
+
+        var result = new string?[handlerMethods.Length * 5];
+
+        var index = 0;
+        foreach (var (methodName, language, requestTypeRef, responseTypeRef, requestContextTypeRef) in handlerMethods)
```

Comment:
I don't know exactly how this works, but wondering if we may have mef cache issues here.  For example if the type used in a handler changes assembly version or moves, does the cache get invalidated?  I guess we would have the same issue now since we're using the Type instance in the metadata?

Reply:
That's right. I'm relying on the VS MEF catalog cache being accurate and invalidated correctly when an assembly changes. Otherwise, I think what you described would have always been a problem, as you also surmised. I think the new potential break with this change would be if the assembly version doesn't change and a method handler changes its request/response/context type. However, I think that's probably OK because that would require a change to the generic arguments of the implemented `IRequestHandler<>` or `INotificationHandler<>`. I would expect such a change to cause the cache to be invalidated.